Experiment with `runAnimations` utility pattern

nickschot opened the pull request:
This cleans up the orchestration flow a little bit. What we're aiming for will likely not be this but at least we're batching the actual web animations now to play together. Theoretically, in the old situation, if a `Sprite.startAnimation()` call takes more than a frame the animations may be out of sync. This prevents that.

nickschot committed:
Add runAnimations util

nickschot committed:
Make Sprite/SpriteAnimation compatible with runAnimations utility

nickschot committed:
Migrate magic-move transition to runAnimations utility

nickschot committed:
Migrate fade transition to runAnimations utility

nickschot committed:
Update interruption demo to runAnimations utility

nickschot committed:
Update routes demo to runAnimations utility

nickschot committed:
Update motion-study demo to runAnimations utility

lukemelia commented:
With this API, how would you handle a situation where you want to start sprite A and sprite B animating at the same time but you want sprite B to animate for 2000ms and sprite A to animate for 1000ms, and once sprite A's animation completes, initiate a 1500ms animation of sprite C?

nickschot commented:
We don't yet! This is sort of an in between step as it now very clearly highlights what problems we need to solve (your example being one of them). If you run the demo and go from the detail page to the index page you can see a bunch of issues that would need to be resolved by any API we design. More tomorrow...

nickschot committed:
Refactor intermediate sprite handling to live on the Changeset

nickschot committed:
Add notes about interruptedSprites

nickschot committed:
Update magic-move and fade transitions to take behavior and duration arguments

nickschot committed:
Add assertion to Sprite.setupAnimation for expected arguments for certain behaviors

nickschot committed:
Update interruption, routes and motion-study demo

nickschot committed:
Let runAnimations take a sprite set instead of a changeset

nickschot committed:
Update motion-study demo

nickschot commented:
I've updated this to make what you state possible in a more straightforward way by passing the sprites rather than the changeset to `runAnimations`. 

Also fixed up the motion-study demo to work correctly in all cases w/ interruption. Nicely highlights what kind of "features" we'll need to provide/make more straightforward to use. Main example being a more ergonomic way of filtering/checking if a sprite was interrupted as there may be unexpected matching sprites during an interruption (.e.g when going from a list to a single item -> interrupt -> suddenly all list items are keptSprites with the same role). For this demo I'm manually doing it by checking if there's a `boundsDelta` https://github.com/cardstack/animations-experiment/pull/23/files#diff-6fc5493e9ed188b40ea534176252b6d029fb8336a21100dbb6802aab16e0f8ffR62

I'm working on an initial version of a "DSL" with the features we discussed. I doubt the end result will look exactly like it, but it does provide an easy way to verify what kind of hooks/config options we'll need and how to provide them from a lower level.

nickschot committed:
Clear sprite's motions after compiling them

nickschot committed:
Update interruption demo to be a little more interesting
